batch/batch.go

```diff
@@ -0,0 +1,279 @@
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util"
+	"golang.org/x/net/context"
+
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+// UpdateForBatch updates the first argument (the header of a request contained
+// in a batch) from the second one (the batch header), returning an error when
+// inconsistencies are found.
+// It is checked that the individual call does not have a UserPriority
+// or Txn set that differs from the batch's.
+// TODO(tschottdorf): will go with #2143.
+func UpdateForBatch(args proto.Request, bHeader proto.RequestHeader) error {
+	// Disallow transaction, user and priority on individual calls, unless
+	// equal.
+	aHeader := args.Header()
+	if aPrio := aHeader.GetUserPriority(); aPrio != proto.Default_RequestHeader_UserPriority && aPrio != bHeader.GetUserPriority() {
+		return util.Error("conflicting user priority on call in batch")
+	}
+	aHeader.UserPriority = bHeader.UserPriority
+	aHeader.Txn = bHeader.Txn // reqs always take Txn from batch
+	return nil
+}
+
+// MaybeWrap wraps the given argument in a batch, unless it is already one.
+// TODO(tschottdorf): will go when proto.Call does.
+func MaybeWrap(args proto.Request) (*proto.BatchRequest, func(proto.Response) proto.Response) {
+	if bArgs, ok := args.(*proto.BatchRequest); ok {
+		return bArgs, func(a proto.Response) proto.Response { return a }
+	}
+	bArgs := &proto.BatchRequest{}
+	bArgs.RequestHeader = *(gogoproto.Clone(args.Header()).(*proto.RequestHeader))
+	bArgs.Add(args)
+	return bArgs, func(reply proto.Response) proto.Response {
+		bReply, ok := reply.(*proto.BatchResponse)
+		if !ok {
+			// Request likely never sent, but caught a local error.
+			return reply
+		}
+		var unwrappedReply proto.Response
+		if len(bReply.Responses) == 0 {
+			unwrappedReply = args.CreateReply()
+		} else {
+			unwrappedReply = bReply.Responses[0].GetValue().(proto.Response)
+		}
+		// The ReplyTxn is propagated from one response to the next request,
+		// and we adopt the mechanism that whenever the Txn changes, it needs
+		// to be set in the reply, for example to ratched up the transaction
+		// timestamp on writes when necessary.
+		// This is internally necessary to sequentially execute the batch,
+		// so it makes some sense to take the burden of updating the Txn
+		// from TxnCoordSender - it will only need to act on retries/aborts
+		// in the future.
+		unwrappedReply.Header().Txn = bReply.Txn
+		return unwrappedReply
+	}
+}
+
+// MaybeWrapCall returns a new call which wraps the original Args and Reply
+// in a batch, if necessary.
+// TODO(tschottdorf): will go when proto.Call does.
+func MaybeWrapCall(call proto.Call) (proto.Call, func(proto.Call) proto.Call) {
+	var unwrap func(proto.Response) proto.Response
+	call.Args, unwrap = MaybeWrap(call.Args)
+	newUnwrap := func(origReply proto.Response) func(proto.Call) proto.Call {
+		return func(newCall proto.Call) proto.Call {
+			origReply.Reset()
```

Comment:
If Merge is a deep copy then this makes sense. You're right, this isn't incorrect, just slightly inefficient compared to the reflect call.